agent/src/main/java/com/spotify/ffwd/statistics/SemanticCoreStatistics.java

```diff
@@ -246,24 +247,36 @@ public void reportMetricsDroppedByFilter(final int dropped) {
     @Override
     public HighFrequencyDetectorStatistics newHighFrequency() {
         final MetricId m = metric.tagged("component", "high-freq");
-        final AtomicLong highFreqMetrics = new AtomicLong();
+        final Map<MetricId, Long> highFreqMetricsMap = new ConcurrentHashMap<>();
 
         return new HighFrequencyDetectorStatistics() {
             private final Counter dropped = registry.counter(
                 m.tagged("what", "dropped-metrics", "unit", "metric"));
 
-            private final Gauge metricsCardinalityMetric =
-              registry.register(m.tagged("what", "high-freq-metrics"),
-                  (Gauge<Long>) () -> (long) highFreqMetrics.get());
-
             @Override
             public void reportHighFrequencyMetricsDropped(int dropped) {
                 this.dropped.inc(dropped);
             }
 
             @Override
-            public void reportHighFrequencyMetrics(int marked) {
-                highFreqMetrics.set(marked);
+            public void reportHighFrequencyMetrics(int marked, String... tags) {
+                MetricId gaugeMetric = m.tagged("what", "high-freq-metrics").tagged(tags);
+
+                registerGauge(gaugeMetric);
+
+                highFreqMetricsMap.put(gaugeMetric, (long) marked);
+            }
+
+            private synchronized void registerGauge(MetricId gauge) {
```

Comment:
synchronizing a read access to a concurrent concurrent hashMap is off it defies the point of using it.

Reply:
@ao2017 @smstone  I'm using the synchronized access here as I'd like to make sure the registration of a gauge won't produce an exception. If there race condition it's possible `highFreqMetricsMap.containsKey` returns false but by the time it tries to register new Gauge other thread already done it. My understanding of synchronized method/block that it won't allow this race condition to happen.

Reply:
Check register.register implementation  you don't need synchronized. 
 I also  prefer the option of handling exception instead of  adding synchronized with a check.

Synchronized only provides serialized access to  this method any thread can change the register outside of  this method. So in this case if you rely on synchronized for thread safety it can easily be broken.

Reply:
I've added try/catch to handle race condition.